service/frontend/wrappers/ratelimited/ratelimit.go

```diff
@@ -36,17 +38,39 @@ const (
 	ratelimitTypeAsync
 )
 
-func (h *apiHandler) allowDomain(requestType ratelimitType, domain string) bool {
+func (h *apiHandler) allowDomain(ctx context.Context, requestType ratelimitType, domain string) (bool, error) {
+	info := quotas.Info{Domain: domain}
+	var policy quotas.Policy
 	switch requestType {
 	case ratelimitTypeUser:
-		return h.userRateLimiter.Allow(quotas.Info{Domain: domain})
+		policy = h.userRateLimiter
 	case ratelimitTypeWorker:
-		return h.workerRateLimiter.Allow(quotas.Info{Domain: domain})
+		policy = h.workerRateLimiter
 	case ratelimitTypeVisibility:
-		return h.visibilityRateLimiter.Allow(quotas.Info{Domain: domain})
+		policy = h.visibilityRateLimiter
 	case ratelimitTypeAsync:
-		return h.asyncRateLimiter.Allow(quotas.Info{Domain: domain})
+		policy = h.asyncRateLimiter
 	default:
 		panic("coding error, unrecognized request ratelimit type value")
 	}
+	
+	// If context has a deadline, use Wait() to potentially wait for a token
+	// Otherwise, use Allow() for an immediate check
+	if _, hasDeadline := ctx.Deadline(); hasDeadline {
```

Comment:
The incoming context likely always has a deadline imposed by the RPC mechanism (gRPC or TChannel), so I think this would apply the wait time to more than we'd like.

I think we want to instead apply it based on the ratelimitType:
```
if requestType == ratelimitTypeWorkerPoll {
  waitTime := h.maxWorkerPollWait(domain)
  if waitTime > 0 {
      return h.waitForPolicy(ctx, waitTime, policy)
  }
}
if !policy.Allow(info) {
  return ErrRateLimited
}
return nil
```

where `waitForPolicy` is a new method that includes the logic you have here along with the context creation logic that's currently in the template.

and `ErrRateLimited` is just a global variable equal to `&types.ServiceBusyError{Message: "Too many outstanding requests to the cadence service"}`.